tensorflow/compiler/mlir/xla/transforms/hlo_outline_fusion.cc

```diff
@@ -0,0 +1,555 @@
+    int from = cluster_from->cycles_graph_node_id();
+    int to = cluster_to->cycles_graph_node_id();
+
+    if (!cycle_detector_.ContractEdge(from, to)) {
+      llvm::dbgs() << "Could not contract " << from << " -> " << to
+                   << " because contracting the edge would create a cycle.";
+      return false;
+    }
+
+    // Merge the clusters.
+    cluster_from->Merge(cluster_to);
+
+    // Merge the UnionFind<Cluster*>.
+    cluster_for_node_[from].Merge(&cluster_for_node_[to]);
+
+    return true;
+  }
+
+  template <typename FnTy>
+  bool ForEachEdgeInPostOrder(FnTy fn) {
+    bool changed = false;
+    for (int32 node : cycle_detector_.AllNodesInPostOrder()) {
+      Cluster* cluster_from = GetClusterForCyclesGraphNode(node);
+      if (!cluster_from) {
+        continue;
+      }
+
+      // Make a copy of the set of successors because we may modify the graph in
+      // TryToContractEdge.
+      std::vector<int32> successors_copy =
+          cycle_detector_.SuccessorsCopy(cluster_from->cycles_graph_node_id());
+
+      for (int to : successors_copy) {
+        Cluster* cluster_to = GetClusterForCyclesGraphNode(to);
+        if (!cluster_to) {
+          continue;
+        }
+
+        bool contracted_edge = fn(cluster_from, cluster_to);
+        changed |= contracted_edge;
+      }
+    }
+
+    return changed;
+  }
+
+  std::vector<Value> GetResultsOfFusedPattern(Cluster* from, Cluster* to) {
+    auto fused_pattern = MergeFusionPattern(
+        from->fused_pattern(), to->fused_pattern());
+    return GetOutputsOfFusionPattern(fused_pattern);
+  }
+
+  bool TryToContractEdge(Cluster* from, Cluster* to) {
+    int node_to = to->cycles_graph_node_id();
+    int node_from = from->cycles_graph_node_id();
+
+    // Both node_to and node_from should be fusible
+    if (!IsFusible(op_list_[node_to]) || !IsFusible(op_list_[node_from])) {
+      return false;
+    }
+
+    // ReduceOp is fusible only when it's the output of the fusion pattern.
+    if (dyn_cast<xla_hlo::ReduceOp>(op_list_[node_from])) {
+      return false;
+    }
+
+    // Only Try to fuse along data edge, not index edge
+    if (IsShapeProducer(op_list_[node_from])) {
+      return false;
+    }
+
+    // All outputs of a kLoop fusion should be same.
+    auto results = GetResultsOfFusedPattern(from, to);
+    auto get_workload_shape = [] (Value v) {
+      if (dyn_cast<xla_hlo::ReduceOp>(v.getDefiningOp())) {
+        v = v.getDefiningOp()->getOperand(0);
+      }
+      return v;
+    };
+    if (!llvm::all_of(results, [&] (Value result) {
+      Value ref = get_workload_shape(results[0]);
+      Value val = get_workload_shape(result);
+      return shape_analysis_.HasSameShape(ref, val);
+    })) {
+      return false;
+    }
+
+    return MergeClusters(from, to);
+  }
+
+  bool RunEdgeContractionLoop() {
+    bool changed = false;
+    changed = ForEachEdgeInPostOrder([&](Cluster* from, Cluster* to) -> bool {
```

Comment:
Why the extra lambda here?

Reply:
`TryToContractEdge` is a member function, thus can not be passed to `ForEachEdgeInPostOrder` directly. I have changed to use `std::bind` instead.